File: ripperdoc/wetware/mode-example/scripts/inject_mode.py
#!/usr/bin/env python3
"""Mode-routing hook — inject the active mode's md file into the model's context.

Wired inline in each host's manifest:
  .claude-plugin/plugin.json   UserPromptSubmit   — Claude
  .codex-plugin/plugin.json    UserPromptSubmit   — Codex
  gemini-extension.json        BeforeAgent        — Gemini

Two lookups, both OS-agnostic (pure pathlib):
  1. preemdeck.json — found by walking up from this script's dir until it turns
     up. Works in-repo and once installed: every host nests the plugin copy
     under its clone root (`~/.claude`, `~/.codex`, `~/.gemini` — see boot.sh),
     and preemdeck.json sits at that root, so it is always an ancestor.
  2. modes/<mode>.md — shipped inside the plugin, resolved relative to this
     script (plugin_root/modes/<mode>.md).

The active mode is the `mode-example` field of preemdeck.json; the matching
modes/<mode>.md body is injected as additionalContext.

  --event <name>  fallback event name when stdin omits `hook_event_name`
                  (default UserPromptSubmit). stdin wins when it carries one.

Always exits 0 — a missing config, missing/empty field, or unknown mode (no
matching md) is a silent no-op (prints `{}`); the host never blocks.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

CONFIG_NAME = "preemdeck.json"
MODE_KEY = "mode-example"
DEFAULT_EVENT = "UserPromptSubmit"
SEARCH_START = Path(__file__).resolve().parent
MODES_DIR = Path(__file__).resolve().parents[1] / "modes"


def find_config(start: Path) -> Path | None:
    """Walk up from `start` (inclusive) toward the filesystem root; first hit wins."""
    for parent in (start, *start.parents):
        candidate = parent / CONFIG_NAME
        if candidate.is_file():
            return candidate
    return None


def select_mode(config: Path) -> str | None:
    """Return the active mode name from the config's `MODE_KEY` field, or None."""
    try:
        data = json.loads(config.read_text(encoding="utf-8"))
    except (json.JSONDecodeError, OSError):
        return None
    if isinstance(data, dict):
        mode = data.get(MODE_KEY)
        if isinstance(mode, str) and mode:
            return mode
    return None


def load_mode_text(mode: str) -> str | None:
    """Load `modes/<mode>.md` from the plugin; None if unknown, empty, or unsafe.

    `mode` must be a bare filename stem — anything carrying a path separator or a
    dot-segment is rejected, so a config value can't escape the modes dir.
    """
    if Path(mode).name != mode:
        return None
    md = MODES_DIR / f"{mode}.md"
    if not md.is_file():
        return None
    return md.read_text(encoding="utf-8").strip() or None


def extract_event(argv: list[str]) -> str | None:
    """Return the value following the first `--event` flag, or None."""
    for i, arg in enumerate(argv):
        if arg == "--event" and i + 1 < len(argv):
            return argv[i + 1]
    return None


def main() -> int:
    config = find_config(SEARCH_START)
    if config is None:
        print("{}")
        return 0
    mode = select_mode(config)
    if mode is None:
        print("{}")
        return 0
    text = load_mode_text(mode)
    if text is None:
        print("{}")
        return 0

    event_name = extract_event(sys.argv[1:]) or DEFAULT_EVENT
    try:
        payload = json.loads(sys.stdin.read() or "{}")
        if isinstance(payload, dict):
            name = payload.get("hook_event_name")
            if isinstance(name, str) and name:
                event_name = name
    except (json.JSONDecodeError, OSError):
        pass

    print(
        json.dumps(
            {
                "hookSpecificOutput": {
                    "hookEventName": event_name,
                    "additionalContext": text,
                }
            }
        )
    )
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: source/ripperdoc/wetware/directive/scripts/inject-mode.ts
#!/usr/bin/env -S preemdeck-runtime
/**
 * inject-mode.ts — directive-routing hook.
 *
 * Walks up from the script dir to find preemdeck.json, reads its `directive`
 * object (slot -> value; a bare string is a single legacy value), resolves each
 * active value to `skills/<value>/directive.md`, and injects the concatenated
 * (slot order, deduped) bodies via lib/hook.ts. A missing config / empty directive
 * / all-unknown values is a silent `{}` no-op. `--event <name>` is the required
 * host event; stdin's hook_event_name wins.
 *
 * Path resolution: SKILLS_DIR = <script-dir>/../skills.
 */

import { existsSync } from "node:fs"
import { readFile, stat } from "node:fs/promises"
import { dirname, join } from "node:path"
import argvex from "argvex"
import { runInjectionHook } from "../../../../common/hook-inject"

const CONFIG_NAME = "preemdeck.json"
const DIRECTIVE_KEY = "directive"

const SEARCH_START = import.meta.dir
const SKILLS_DIR = join(dirname(import.meta.dir), "skills")

/** Walk up from `start` (inclusive) toward the root; first preemdeck.json wins. */
export const findConfig = async (start: string): Promise<string | null> => {
    let dir = start
    for (;;) {
        const candidate = join(dir, CONFIG_NAME)
        if (existsSync(candidate) && (await stat(candidate)).isFile()) return candidate
        const parent = dirname(dir)
        if (parent === dir) return null
        dir = parent
    }
}

/** Active values from the config's `directive` field, in slot order, deduped. */
export const selectVariants = async (config: string): Promise<string[]> => {
    let data: unknown
    try {
        data = JSON.parse(await readFile(config, "utf8"))
    } catch {
        return []
    }
    if (data === null || typeof data !== "object" || Array.isArray(data)) return []
    const field = (data as Record<string, unknown>)[DIRECTIVE_KEY]
    let values: unknown[]
    if (typeof field === "string") {
        values = [field]
    } else if (field !== null && typeof field === "object" && !Array.isArray(field)) {
        values = Object.values(field as Record<string, unknown>)
    } else {
        return []
    }
    const out: string[] = []
    for (const v of values) {
        if (typeof v === "string" && v && !out.includes(v)) out.push(v)
    }
    return out
}

/**
 * Load `skills/<value>/directive.md`; null if unknown, empty, or unsafe.
 * `value` must be a bare name (no path separator / dot-segment) so a config value
 * can't escape the skills dir.
 */
export const loadModeText = async (skillsDir: string, value: string): Promise<string | null> => {
    if (value.includes("/") || value === ".") return null
    const body = join(skillsDir, value, "directive.md")
    if (!existsSync(body) || !(await stat(body)).isFile()) return null
    const text = (await readFile(body, "utf8")).trim()
    return text || null
}

/** Value of the first/only `--event`, or null. Never throws. */
export const extractEvent = (argv: string[]): string | null => {
    try {
        return argvex({ argv, schema: [{ name: "event", arity: 1 }] }).event?.[0] ?? null
    } catch {
        return null
    }
}

/** Build the concatenated directive bodies for the active config, or "" / null. */
export const renderBodies = async (searchStart: string, skillsDir: string): Promise<string | null> => {
    const config = await findConfig(searchStart)
    if (config === null) return null
    const bodies: string[] = []
    for (const v of await selectVariants(config)) {
        const t = await loadModeText(skillsDir, v)
        if (t) bodies.push(t)
    }
    if (bodies.length === 0) return null
    return bodies.join("\n\n")
}

if (import.meta.main) {
    const cliEvent = extractEvent(Bun.argv.slice(2))
    if (cliEvent === null || cliEvent.length === 0) {
        process.stderr.write("usage: inject-mode --event <name>\n")
        process.exit(2)
    }
    const bodies = await renderBodies(SEARCH_START, SKILLS_DIR)
    await runInjectionHook({
        event: cliEvent,
        render: () => bodies
    })
    process.exit(0)
}
